### PR TITLE
Add Model Definition

### PR DESCRIPTION
**Pull request by MEichhoff**
example generated model definition: 
https://gist.github.com/MEichhoff/9ad8ea5a3ae279adb757c83ebcee9d0c

### CODE REVIEW

File: src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/SourceAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+        public string SourceUrl
+        {
+            get => sourceUrl;
+        }
+
+        public string Endpoint
+        {
+            get => endpoint;
+        }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/SourceAttribute.cs:23]: I would suggest transform them in one-line statements, like:

```suggestion
        public string SourceUrl => sourceUrl;

        public string Endpoint => endpoint;
```

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+{
+    public static class ModelDefinitionService
+    {
+        public static ModelDefinition GenerateDefinitionForController(Type controller)
+        {
+            var modelDefinition = new ModelDefinition();
+
+            var methods = controller.GetMethods();
+
+            var getMethod = GetOperationFromAttribute(methods, typeof(ModelDefinitionGetOperationAttribute));
```
[the remark this answers is not in the document]
Comment (simplicbe) [reply src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:15]: But I would only start with this type imho.

File: src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ModelDefinitionPostOperationAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+            this.endpoint = endpoint;
+            this.response = response;
+            this.request = request;
+        }
+
+        public string Endpoint { get => endpoint; }
+
+        public Type Request { get => request; }
+
+        public Type Response { get => response; }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ModelDefinitionPostOperationAttribute.cs:21]:
```suggestion
        public string Endpoint => endpoint;

        public Type Request => request;

        public Type Response => response;
```

File: src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ModelDefinitionPatchOperationAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+            this.endpoint = endpoint;
+            this.response = response;
+            this.request = request;
+        }
+
+        public string Endpoint { get => endpoint; }
+
+        public Type Request { get => request; }
+
+        public Type Response { get => response; }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ModelDefinitionPatchOperationAttribute.cs:21]:
```suggestion
        public string Endpoint => endpoint;

        public Type Request => request;

        public Type Response => response;
```

File: src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ModelDefinitionGetOperationAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+        public ModelDefinitionGetOperationAttribute(string endpoint, Type response)
+        {
+            this.endpoint = endpoint;
+            this.response = response;
+        }
+
+        public string Endpoint { get => endpoint; }
+
+        public Type Response { get => response; }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ModelDefinitionGetOperationAttribute.cs:17]:
```suggestion
        public string Endpoint => endpoint;

        public Type Response => response;
```

File: src/Simplic.OxS.ModelDefinition.Service/TypeExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+            if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IEnumerable<>))
+            {
+                return true;
+            }
+
+            if (typeof(IEnumerable).IsAssignableFrom(type))
+            {
+                return true;
+            }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Service/TypeExtensions.cs:42]: Don't use braces in this if-statements.

File: src/Simplic.OxS.ModelDefinition/ModelDefinition.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        public string? Model { get; set; }
+
+        public string SourceUrl { get; set; }
+
+        public Operations Operations { get; set; }
+            = new Operations();
+
+        public IList<DataSource> DataSources { get; set; }
+
+        public IList<PropertyDefinition> Properties { get; set; }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition/ModelDefinition.cs:22]: Why don't you initialize this lists, but the one below? If they should not be initialized, use `?`

File: src/Simplic.OxS.ModelDefinitionExtension.Test/TestEnv/TestEnum.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Simplic.OxS.ModelDefinitionExtension.Test.TestEnv
+{
+    internal enum TestEnum: int
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinitionExtension.Test/TestEnv/TestEnum.cs:9]:
```suggestion
    internal enum TestEnum : int
```

File: src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/AvailableTypeAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+        public AvailableTypeAttribute(Type type)
+        {
+            this.type = type;
+        }
+
+        public Type Type
+        {
+            get { return type; }
+        }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/AvailableTypeAttribute.cs:16]:
```suggestion
        public Type Type => type;
```

File: src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ModelDefinitionDeleteOperationAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    public sealed class ModelDefinitionDeleteOperationAttribute : Attribute
+    {
+        private readonly string endpoint;
+
+        public ModelDefinitionDeleteOperationAttribute(string endpoint)
+        {
+            this.endpoint = endpoint;
+        }
+
+        public string Endpoint { get => endpoint; }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ModelDefinitionDeleteOperationAttribute.cs:13]:
```suggestion
        public string Endpoint => endpoint;
```

File: src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ReferenceIdAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    public sealed class ReferenceIdAttribute : Attribute
+    {
+        private readonly string referenceIdPropertyName;
+
+        public ReferenceIdAttribute(string referenceIdPropertyName)
+        {
+            this.referenceIdPropertyName = referenceIdPropertyName;
+        }
+
+        public string ReferenceIdPropertyName { get { return this.referenceIdPropertyName; } }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Extenstion.Abstractions/ReferenceIdAttribute.cs:19]: Any reason for `this.`?

```suggestion
        public string ReferenceIdPropertyName => referenceIdPropertyName;
```

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+                Endpoint = deleteAttribute.Endpoint,
+                Type = "http-delete"
+            };
+        }
+
+        private static void BuildPropertiesOrReference(ModelDefinition modelDefinition, Type response)
+        {
+            if (modelDefinition.Model != null
+                &&
+                modelDefinition.Model.Equals(GetReferenceName(response)))
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:119]:
```suggestion
            if (modelDefinition.Model != null
                && modelDefinition.Model.Equals(GetReferenceName(response)))
```

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+                    AddRefereceDefinition(modelDefinition, property.PropertyType);
+                }
+
+                properties.Add(propertyDefinition);
+            }
+
+            return properties;
+        }
+
+        private static (string?, string?) GetMinAndMaxValue(PropertyInfo property)
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:243]: I would suggest working with named tuples:

Like here with min and max: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples#use-cases-of-tuples

```suggestion
        private static (string? min, string? max) GetMinAndMaxValue(PropertyInfo property)
```

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+                referenceDefinition.SourceUrl = modelDefinition.SourceUrl;
+            }
+
+            modelDefinition.References.Add(referenceDefinition);
+        }
+
+        private static string GetReferenceName(Type type)
+        {
+            return $"${type.Name}";
+        }
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:361]:
```suggestion
        private static string GetReferenceName(Type type) => $"${type.Name}";
```

**Review by simplicbe — CHANGES_REQUESTED** [review]
A few small changes, in general it looks pretty good

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+                Type = "http-post"
+            };
+
+            BuildPropertiesOrReference(modelDefinition, postAttribute.Request);
+            BuildPropertiesOrReference(modelDefinition, postAttribute.Response);
+        }
+
+        private static void BuildPatch(ModelDefinition modelDefinition, MethodInfo patchMethod)
+        {
+
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:83]:
```suggestion
```

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+            if (property.PropertyType == typeof(string))
+            {
+
+                var stringLenthAttribute = Attribute.GetCustomAttribute(property, typeof(StringLengthAttribute))
+                    as StringLengthAttribute;
+
+                if (stringLenthAttribute != null)
+                    return (stringLenthAttribute.MinimumLength.ToString(), stringLenthAttribute.MaximumLength.ToString());
+
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:256]:
```suggestion
```

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+        }
+
+        #endregion
+
+        #region Properties
+
+        private static IList<PropertyDefinition> BuildProperties(Type model, ModelDefinition modelDefinition)
+        {
+            var properties = new List<PropertyDefinition>();
+            foreach (var property in model.GetProperties())
```
Comment (simplicbe) [inline src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:141]: *Important*

Here you are reading all properties from the model. Does this support inheritance. For example:

```cs
public class ModelA 
{
    public string Name { get; set; }
}

public class ModelB : ModelA
{
    public string Street { get; set; }
}

```

if properties of `ModelB` are read?

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MEichhoff) [reply src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:15]: We could argue about adding put, but others should in my optinion not be covered in the model definition

[the remark this answers is not in the document]
Comment (simplicbe) [reply src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:15]: If adding "Put" is not a big deal, I think that would be great.

File: src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs (hunk 17 above)
Comment (MEichhoff) [reply src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:141]: Yes, in the given example e.g the create tour request contains just a collection of actions itself, other properties come from an inherited class and all of the properties are  present in the model definition.

Comment (simplicbe) [reply src/Simplic.OxS.ModelDefinition.Service/ModelDefinitionService.cs:141]: Perfect

**Pull request merged**